src/codewhisperer/service/keyStrokeHandler.ts

```diff
@@ -215,3 +158,113 @@ export class KeyStrokeHandler {
+        const tabSize = getTabSizeSetting()
+        if (str.length % tabSize === 0 && str.trim() === '') {
+            return true
+        }
+        return false
+    }
+
+    static isUserTypingSpecialChar(str: string): boolean {
+        const specialChars = new Map<string, string>([
+            ['(', ')'],
+            ['[', ']'],
+            ['{', '}'],
+            [':', ':'],
+        ])
+        if (specialChars.has(str[0])) {
+            const substr = str.substring(1)
+            // TODO: improve here with matching closing brackets
+            return substr.length === 0 || substr.length === 1
+        }
+
+        return false
+    }
+
+    static isSingleLine(str: string): boolean {
+        let newLineCounts = 0
+        for (const ch of str) {
+            if (ch === '\n') newLineCounts += 1
+        }
+
+        // since pressing Enter key possibly will generate string like '\n        ' due to indention
+        if (this.isEnterKey(str)) return true
+        if (newLineCounts >= 1) return false
+        return true
+    }
+
+    static isFormatting(str: string): boolean {
+        const rExp: RegExp = /^[\s]+$/g
+        return rExp.test(str)
+    }
+}
+
+// Case when event.contentChanges.length === 1
+export class SingleChange extends DocumentChangedType {
+    constructor(contentChanges: ReadonlyArray<vscode.TextDocumentContentChangeEvent>) {
+        super(contentChanges)
+    }
+
+    override checkChangeSource(): DocumentChangedSource {
+        const changedText = this.contentChanges[0].text
+
+        if (DocumentChangedType.isSingleLine(changedText)) {
+            if (changedText === '') {
+                return DocumentChangedSource.Deletion
+            } else if (DocumentChangedType.isEnterKey(changedText)) {
+                return DocumentChangedSource.EnterKey
+            } else if (DocumentChangedType.isTabKey(changedText)) {
+                return DocumentChangedSource.TabKey
+            } else if (DocumentChangedType.isUserTypingSpecialChar(changedText)) {
+                return DocumentChangedSource.SpecialCharsKey
+            } else if (changedText.length === 1) {
+                return DocumentChangedSource.RegularKey
+            } else if (DocumentChangedType.isFormatting(changedText)) {
+                // this logic must be checked after EnterKey
+                return DocumentChangedSource.Reformatting
+            } else {
+                return DocumentChangedSource.IntelliSense
+            }
+        }
+
+        // Won't trigger cwspr on multi-line changes
+        return DocumentChangedSource.Unknown
+    }
+}
+
+// Case when event.contentChanges.length > 1
+class MultiChange extends DocumentChangedType {
+    constructor(contentChanges: ReadonlyArray<vscode.TextDocumentContentChangeEvent>) {
+        super(contentChanges)
+    }
+
+    override checkChangeSource(): DocumentChangedSource {
+        return DocumentChangedSource.Reformatting
+    }
+}
```

Comment:
For now, the only case I observed `events.contentChanges.length > 1` is when formatter like prettier is working.
Cases like pressing keyboard '(', '{' etc. and IDE generating the closing ones are still of length === 1

Reply:
Parenthesis case:
https://user-images.githubusercontent.com/96078566/191853703-7b41ada5-00ac-4b04-b32f-26e4ecac9240.mov